apps/glusterfs/app_device.go

```diff
@@ -322,35 +322,31 @@ func (a *App) DeviceSetState(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, "request unable to be parsed", 422)
 		return
 	}
+	logger.Info("entered set state state is %+v, device is %v", msg, id)
 
 	// Set state
-	err = a.db.Update(func(tx *bolt.Tx) error {
-		device, err := NewDeviceEntryFromId(tx, id)
+	a.asyncManager.AsyncHttpRedirectFunc(w, r, func() (string, error) {
+		err = a.db.Update(func(tx *bolt.Tx) error {
+			device, err := NewDeviceEntryFromId(tx, id)
+			if err != nil {
+				logger.LogError("Did not find device requested %v", id)
+				return err
+			}
+
+			// Set state
+			err = device.SetState(tx, a.executor, a.allocator, msg.State)
```

Comment:
Transferring the `tx` instead of the `db` here is causing concurrency issues.  For example, this change introduces locks and unlocks in the concurrent brick creation were there were none before.  It looks to me that the Update() call can chage the state first in the `tx` context, then the db context can be freed.  This will let the bricks all be executed in parallel, which is one of the benefits of Heketi.  By introducing locks (I haven't checked yet), but it may create an issue on many nodes and volumes because there will no longer be parallelism.